File: algorithms/[19]删除链表的倒数第N个节点/solution.py
# topics = ["链表", "双指针"]

from __future__ import annotations
from typing import Union


class ListNode:
    def __init__(self, x: int = 0):
        self.val = x
        self.next: Union[ListNode, None] = None


class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        start = ListNode()
        start.next = head

        i = j = start
        while n:
            j = j.next
            n -= 1
        while j.next:
            i = i.next
            j = j.next

        i.next = i.next.next
        return start.next

File: algorithms/[19]删除链表的倒数第N个节点/Solution.cs
// topics = ["链表", "双指针"]

public class ListNode
{
    public int val;
    public ListNode next;
    public ListNode(int val = 0, ListNode next = null)
    {
        this.val = val;
        this.next = next;
    }
}


public class Solution
{
    public ListNode RemoveNthFromEnd(ListNode head, int n)
    {
        ListNode dummy = new ListNode();
        dummy.next = head;

        ListNode i = dummy;
        ListNode j = dummy;

        while (n > 0)
        {
            j = j.next;
            n--;
        }

        while (j.next != null)
        {
            i = i.next;
            j = j.next;
        }

        i.next = i.next.next;

        return dummy.next;
    }
}
